src/paths/floyd_warshall.c

```diff
@@ -140,3 +142,132 @@ igraph_error_t igraph_distances_floyd_warshall(
+        igraph_real_t w = weights ? VECTOR(*weights)[e] : 1;
+
+        if (w < 0) {
+            if (mode == IGRAPH_ALL) {
+                IGRAPH_ERRORF("Negative edge weight (%g) found in undirected graph "
+                              "while calculating distances with Floyd-Warshall.",
+                              IGRAPH_ENEGLOOP, w);
+            } else if (to == from) {
+                IGRAPH_ERRORF("Self-loop with negative weight (%g) found "
+                              "while calculating distances with Floyd-Warshall.",
+                              IGRAPH_ENEGLOOP, w);
+            }
+        }
+
+        if (out && MATRIX(*res, from, to) > w) MATRIX(*res, from, to) = w;
+        if (in  && MATRIX(*res, to, from) > w) MATRIX(*res, to, from) = w;
+    }
+
+    igraph_matrix_int_t successors;
+    igraph_stack_int_t stack;
+
+    IGRAPH_STACK_INT_INIT_FINALLY(&stack, no_of_nodes);
+
+    IGRAPH_CHECK(igraph_matrix_int_init(&successors, no_of_nodes, no_of_nodes));
+    IGRAPH_FINALLY(igraph_matrix_int_destroy, &successors);
+
+    for (igraph_integer_t v=0; v < no_of_nodes; v++) {
+        for (igraph_integer_t u=0; u < no_of_nodes; u++) {
+            /* the second vertex on the shortest path from u to v */
+            MATRIX(successors, u, v) = v;
+        }
+    }
+
+    igraph_vector_int_t children;
+    IGRAPH_VECTOR_INT_INIT_FINALLY(&children, 0);
+    igraph_vector_int_t parents;
+    IGRAPH_VECTOR_INT_INIT_FINALLY(&parents, no_of_nodes);
+
+    for (igraph_integer_t k=0; k < no_of_nodes; k++) {
+        for (igraph_integer_t v=0; v < no_of_nodes; v++) {
+            VECTOR(parents)[v] = v == k ? -1 : MATRIX(successors, v, k);
+        }
+        /* tree of so far discovered shortest paths ending at k */
+        igraph_t tree_in_k;
+        IGRAPH_FINALLY(igraph_destroy, &tree_in_k);
+        IGRAPH_CHECK(igraph_tree_from_parent_vector(&tree_in_k, &parents, IGRAPH_TREE_OUT));
+         /* Iteration order matters for performance!
+          * First j, then i, because matrices are stored as column-major. */
+        for (igraph_integer_t j=0; j < no_of_nodes; j++) {
+            igraph_real_t dkj = MATRIX(*res, k, j);
+            if (dkj == IGRAPH_INFINITY) continue;
+            IGRAPH_CHECK(igraph_stack_int_push(&stack, k));
+            while (!igraph_stack_int_empty(&stack)) {
+                igraph_integer_t v = igraph_stack_int_pop(&stack);
+                igraph_neighbors(&tree_in_k, &children, v, IGRAPH_OUT);
+                igraph_integer_t no_of_children = igraph_vector_int_size(&children);
+                for (igraph_integer_t l=0; l < no_of_children; l++) {
+                    igraph_integer_t i = VECTOR(children)[l];
+                    igraph_real_t di = MATRIX(*res, i, k) + dkj;
+                    igraph_real_t dd = MATRIX(*res, i, j);
+                    if (di < dd) {
+                        MATRIX(*res, i, j) = di;
+                        MATRIX(successors, i, j) = MATRIX(successors, k, j);
```

Comment:
_With the current notation_, I believe this line should read

```suggestion
                        MATRIX(successors, i, j) = MATRIX(successors, i, k);
```

The role of $i$ and $j$ is reversed here compared to the paper and `successors` is transposed compared to $\pi$.

Reply:
Correct! Fixed.